app/assets/javascripts/listings/ListingDataService.js.coffee

```diff
@@ -0,0 +1,434 @@
+        # kick them out unless there's a real listing
+        return $state.go('dahlia.welcome')
+      else if !Service.isAcceptingOnlineApplications(Service.listing)
+        # kick them back to the listing
+        return $state.go('dahlia.listing', {id: id})
+    ).catch( (response) ->
+      deferred.reject(response)
+    )
+    deferred.promise
+
+  Service.getListingAMI = ->
+    angular.copy([], Service.AMICharts)
+    Service.loading.ami = true
+    Service.error.ami = false
+    # shouldn't happen, but safe to have a guard clause
+    return $q.when() unless Service.listing.chartTypes
+    allChartTypes = _.sortBy(Service.listing.chartTypes, 'percent')
+    data =
+      'year[]': _.map(allChartTypes, 'year')
+      'chartType[]': _.map(allChartTypes, 'chartType')
+      'percent[]': _.map(allChartTypes, 'percent')
+    $http.get('/api/v1/listings/ami.json', { params: data }).success((data, status, headers, config) ->
+      if data && data.ami
+        angular.copy(Service._consolidatedAMICharts(data.ami), Service.AMICharts)
+      Service.loading.ami = false
+    ).error( (data, status, headers, config) ->
+      Service.loading.ami = false
+      Service.error.ami = true
+      return
+    )
+
+  Service._consolidatedAMICharts = (amiData) ->
+    charts = []
+    amiData.forEach (chart) ->
+      # look for an existing chart at the same percentage level
+      amiPercentChart = _.find charts, (c) -> c.percent == chart.percent
+      if !amiPercentChart
+        # only push chart if it has any values
+        charts.push(chart) if chart.values.length
+      else
+        # if it exists, modify it with the max values
+        i = 0
+        amiPercentChart.values.forEach (incomeLevel) ->
+          chartAmount = if chart.values[i] then chart.values[i].amount else 0
+          incomeLevel.amount = Math.max(incomeLevel.amount, chartAmount)
+          i++
+    charts
+
+  Service.priorityTypes = (listing) ->
+    Service.collectTypes(listing, 'prioritiesDescriptor')
+
+  Service.collectTypes = (listing, specialType) ->
+    _.map listing[specialType], (descriptor) ->
+      descriptor.name
+
+  Service.priorityLabel = (priority, modifier) ->
+    labelMap =
+      'Vision impairments':
+        name: 'Vision Impairments'
+        description: 'impaired vision'
+      'Hearing impairments':
+        name: 'Hearing Impairments'
+        description: 'impaired hearing'
+      'Hearing/Vision impairments':
+        name: 'Vision and/or Hearing Impairments'
+        description: 'impaired vision and/or hearing'
+      'Mobility/hearing/vision impairments':
+        name: 'Mobility, Hearing and/or Vision Impairments'
+        description: 'impaired mobility, hearing and/or vision'
+      'Mobility impairments':
+        name: 'Mobility Impairments'
+        description: 'impaired mobility'
+
+    return priority unless labelMap[priority]
+    return labelMap[priority][modifier]
+
+  Service.reservedLabel = (listing, type,  modifier) ->
```

Comment:
this one too, would it make sense to move it to listingConstantsService?

Reply:
ah, that's a good point. you're right that we should be strict about what a constant is. let's not move it